### ISSUE TITLE
calculator should produce sensible results with any subset of input variables

### ISSUE THREAD

**Issue by MattHJensen** [open-source-economics/Tax-Calculator#425]
The calculator should be able to produce sensible results with any subset of input variables, especially when the extrapolation is turned off. Variables that are missing should assumed to be zeros. (Sometimes an entire grouping of variables may need to be assumed to be zero if any one of the variables in the grouping is missing). 
 
I believe satisfying this requirement will require the following changes to both the `Records` class and to `functions.py`: in `Records`, arrays of zeros will need to be created based on the presence of the variable in the data rather than based on whether the variable is listed in `ZEROED_NAMES`. In `functions.py`, logic will need to be controlled by the presence of variables (or presence of all variables in a groups of variables) rather than by the value of `puf` or `cmp`. 

cc @martinholmer, @feenberg, @talumbau, @Amy-Xu

**Comment by MattHJensen** [comment]
This is related to https://github.com/open-source-economics/Tax-Calculator/issues/319#issuecomment-122541546.

**Comment by martinholmer** [comment]
@MattHJensen said:
> The calculator should be able to produce sensible results with any subset of input variables, especially when the extrapolation is turned off. Variables that are missing should assumed to be zeros. (Sometimes an entire grouping of variables may need to be assumed to be zero if any one of the variables in the grouping is missing).
>
> I believe satisfying this requirement will require the following changes to both the Records class and to functions.py: in Records, arrays of zeros will need to be created based on the presence of the variable in the data rather than based on whether the variable is listed in ZEROED_NAMES. In functions.py, logic will need to be controlled by the presence of variables (or presence of all variables in a groups of variables) rather than by the value of puf or cmp.

@Amy-Xu and Sean @GoFroggyRun,
Matt has asked me to coordinate the work involved in resolving this issue #425 along with issues #452 and #436.  In this comment, I describe a strategy for resolving these issues for your comment and discussion.  After reaching a consensus on strategy, we will break the strategy down into smaller pieces of work and assign them to you.  But both Matt and I feel it is important for you to be involved in the strategic discussion.

As background, part of Matt's motivation for raising this issue is that in the coming months we might be working with other organizations that have their own tax models to conduct cross-model validation work.  So, the strategy I describe below tries to move the Tax-Calculator code in a direction that will make that collaboration easier.

What Matt says above about what's involved is true, but we can simplify the strategy by restricting ourselves (for now, at least) to making the Tax-Calculator work with any subset of the data currently in the OSPC-generated IRS-PUF/Census-CPS match file (which we call, perhaps confusingly, ```puf.csv```). If we narrow the scope of our efforts in this way, it seems as if we can avoid (for now) changing anything with respect to the puf and cmp variables in the functions.py code.

As Matt says above, the key to resolving this issue is to allow the variable names in the first row of the input file (now called ```puf.csv```) to determine what is now hardwired in the Records.NAMES list.  It seems we could do this fairly easily if we create a set containing the names of __all__ the variables used in the Tax-Calculator and call it ALL_VARNAMES.  Essentially, this set would combine the content in the current NAMES and ZEROED_NAMES lists.  Then when reading the input data file, we check that each variable name is in ALL_VARNAMES and then add that variable name to a second set called USED_VARNAMES.  After processing all the input variables, we construct UNUSED_VARNAMES as:
```
UNUSED_VARNAMES = ALL_VARNAMES - USED_VARNAMES
```
Then we create zero arrays for each of the variables in UNUSED_VARNAMES (just as is done now).

This simple strategy raises an implementation problem that we need to solve before proceeding with the above revisions to the Records class code.  The problem is that there are a few variable names in the input data file that do not correspond to the variable names used in the Tax-Calculator code.  Perhaps this can be handled in a variety of ways, but in my proposed strategy we adopt a simple solution to this problem.  We eliminate differences between variable names where the elimination goes in the direction of standardizing on IRS-SOI PUF variable names.  This approach would imply the following:

* making the 27 variable names in the NAMES list that are uppercase in the code but lowercase in the input data file's first row, all uppercase in the input data file.  [We could make these names all lowercase in the code, which would be more in line with the eNNNNN variable names and Python convention, but this seems like too much search-and-replace work.]

* making the 2 variable names in the NAMES list that are changed from p to e (87482 and 87521) be pNNNNN variables in the Tax-Calculator code.

* eliminate the aliasing business complete so that this following code is removed and all the eNNNNN variable names in the code are replaced with pNNNNN or sNNNNN variable names.
```
# specify eNNNNN aliases for several pNNNNN and sNNNNN variables
        self.e22250 = self.p22250
        self.e04470 = self.p04470
        self.e23250 = self.p23250
        self.e25470 = self.p25470
        self.e08000 = self.p08000
        self.e60100 = self.p60100
        self.e27860 = self.s27860
```
This third bullet, I believe will not only bring more standardization to the Tax-Calculator's variable naming, but do so in a way that will (a) facilitate cross-model validation work with other organizations familiar with the IRS-SOI PUF data, and (b) solve issue #452.

After we do all of the above, then we will be in a much better position to think about issue #436.

Please think about this strategy.  Am I forgetting something?  Is there a better approach to resolving these issues?  Put your comments here in the coming days.  There is no rush on this because I'm going to be on vacation after Thanksgiving and not back in the office until Tuesday, December 8.  We will not begin coding work until after my return and our determination of a consensus strategy.

**Comment by GoFroggyRun** [comment]
@martinholmer 
I was a little bit confused by the third bullet you mentioned. But if I were not mistaken, your plan is to keep those `p` and `s` variables in `functions.py`, instead of standardizing them, given that they are not part of the IRS-SOI PUF data, it that right?

Regarding your strategy dealing with inputs, I think it's pretty reasonable and straightforward. I do have, however, a few concerns. We currently are having two major groups of variables: namely `NAMES` and `ZEROED_NAMES`. Within those `ZEROED_NAMES`, there are two sub-groups: one is for the purpose of initializing, the other is not available for the time being. `c00100` would be a typical example for the initializing sub-group, while `e03700` can be an example for the other one. Notice the two sub-groups form a partition of  `ZEROED_NAMES`. 

While loading the data, do we want to assume that variables like `c00100` might also be included? If not, then those variables will always be inside the `UNUSED_VARNAMES` list. If that is the case, we could shorten the variables in `ALL_VARNAMES` by eliminating `c00100`-like variables, which should improve the efficiency of processing data. 

So, the question here is that how general we will be allowing our input data to be. If we were to allow variables like `c00100` in the input, then should we avoid any further calculation involving `c00100`, possibly by not referencing that particular function? If so, then how about other outputs in that function. This can be really problematic. 

Thus, I think we should be comfortable enough to rule out those variables that shall never appear in the input file, although in the `ZEROED_NAMES`, like `c00100`.

cc @MattHJensen

**Comment by martinholmer** [comment]
Sean Wang (@GoFroggyRun) said:
> I do have, however, a few concerns. We currently are having two major groups of variables: namely NAMES and ZEROED_NAMES. Within those ZEROED_NAMES, there are two sub-groups: one is for the purpose of initializing, the other is not available for the time being. c00100 would be a typical example for the initializing sub-group, while e03700 can be an example for the other one. Notice the two sub-groups form a partition of ZEROED_NAMES.
>
> While loading the data, do we want to assume that variables like c00100 might also be included? If not, then those variables will always be inside the UNUSED_VARNAMES list. If that is the case, we could shorten the variables in ALL_VARNAMES by eliminating c00100-like variables, which should improve the efficiency of processing data.
>
> So, the question here is that how general we will be allowing our input data to be. If we were to allow variables like c00100 in the input, then should we avoid any further calculation involving c00100, possibly by not referencing that particular function? If so, then how about other outputs in that function. This can be really problematic.

Sean, Thanks for the excellent observation.  You are absolutely correct: my original pseudo code did not guard against reading in variables that are calculated.  Matt asked for the ability to read in a **subset** (not a superset) of the variables currently being read into the Tax-Calculator.

It seems as if we can guard against reading in calculated variables by specifying three sets:
(1) set of ```VALID_READ_VARS``` containing the variable names in the ```puf.csv``` header line.
(2) set of ```CALCULATED_VARS``` containing the variable names now in ZEROED_NAMES.
(3) set of ```READ_VARS``` containing no variable names at the start of the Records read logic.

Then the Records class read logic would be something like this:
```
        # remove the aggregated record from 2009 PUF
        tax_dta = tax_dta[tax_dta.RECID != 999999]
        self.dim = len(tax_dta)
        # create variables using tax_dta DataFrame column names
        for varname in list(tax_dta.columns.values):
            if varname not in Records.VALID_READ_VARS:
                msg = 'Records data variable name {} not in VALID_READ_VARS'
                raise ValueError(msg.format(varname))
            Records.READ_VARS.add(varname)
            setattr(self, varname, tax_dta[varname].values)
        # create variables that are set to all zeros
        Records.UNREAD_VARS = Records.VALID_READ_VARS - Records.READ_VARS
        Records.ZEROED_VARS = Records.CALCULATED_VARS | Records.UNREAD_VARS
        for varname in Records.ZEROED_VARS:
            setattr(self, varname, np.zeros((self.dim,)))
```
Does this revised approach make more sense?

cc @MattHJensen @Amy-Xu

**Comment by feenberg** [comment]
On Tue, 15 Dec 2015, Martin Holmer wrote:

> 
> Sean Wang (@GoFroggyRun) said:
>
>       I do have, however, a few concerns. We currently are having two major groups of variables: namely NAMES and
>       ZEROED_NAMES. Within those ZEROED_NAMES, there are two sub-groups: one is for the purpose of initializing, the other
>       is not available for the time being. c00100 would be a typical example for the initializing sub-group, while e03700
>       can be an example for the other one. Notice the two sub-groups form a partition of ZEROED_NAMES.
>
>       While loading the data, do we want to assume that variables like c00100 might also be included? If not, then those
>       variables will always be inside the UNUSED_VARNAMES list. If that is the case, we could shorten the variables in
>       ALL_VARNAMES by eliminating c00100-like variables, which should improve the efficiency of processing data.
>
>       So, the question here is that how general we will be allowing our input data to be. If we were to allow variables
>       like c00100 in the input, then should we avoid any further calculation involving c00100, possibly by not referencing
>       that particular function? If so, then how about other outputs in that function. This can be really problematic.
> 
> Sean, Thanks for the excellent observation. You are absolutely correct: my original pseudo code did not guard against reading in
> variables that are calculated. Matt asked for the ability to read in a subset (not a superset) of the variables currently being
> read into the Tax-Calculator.
> 
> It seems as if we can guard against reading in calculated variables by specifying three sets:

It is convenient to allow extra variables in the input, so that one can 
aggregate and tab the additional variables without merging each time. 
cariables like C00100 should be ok too, since they are never used before
being set. In the SAS code I try to make sure that every C variable is
set before being used - it does take some attention to keep that up.

It is valuable to let c00100 be in the dataset since that means one can 
recalculate tax without dropping variables.

dan

**Comment by martinholmer** [comment]
Dan @feenberg said:
> It is convenient to allow extra variables in the input, so that one can aggregate and tab the additional variables without merging each time.  C variables like C00100 should be ok too, since they are never used before being set. In the SAS code I try to make sure that every C variable is set before being used - it does take some attention to keep that up.
>
> It is valuable to let c00100 be in the dataset since that means one can recalculate tax without dropping variables.

Dan, Thanks for your comment on this issue.  I'll think about your request, which seems to be a convenience issue (right?).  However, my first impression is that is goes against the main thrust of what Matt is trying to achieve in these changes.  And Sean was worried about allowing C variables to be read in because allowing that would require considerable extra complications to the code.

cc @MattHJensen @GoFroggyRun @Amy-Xu

**Comment by feenberg** [comment]
On Tue, 15 Dec 2015, Martin Holmer wrote:

> 
> Dan @feenberg said:
>
>       It is convenient to allow extra variables in the input, so that one can
>       aggregate and tab the additional variables without merging each time. C
>       variables like C00100 should be ok too, since they are never used before
>       being set. In the SAS code I try to make sure that every C variable is set
>       before being used - it does take some attention to keep that up.
>
>       It is valuable to let c00100 be in the dataset since that means one can
>       recalculate tax without dropping variables.
> 
> Dan, Thanks for your comment on this issue. I'll think about your request, which seems
> to be a convenience issue (right?). However, my first impression is that is goes
> against the main thrust of what Matt is trying to achieve in these changes. And Sean
> was worried about allowing C variables to be read in because allowing that would
> require considerable extra complications to the code.

That surprises me. Can you explain why surplus variables are a problem? 
Are we depending on Python for initialization of C-values? Do extra 
variables require special declarations?

dan

**Comment by martinholmer** [comment]
Dan @feenberg said:
> That surprises me. Can you explain why surplus variables are a problem?  Are we depending on Python for initialization of C-values? Do extra variables require special declarations?

Maybe "surplus variables" are not a problem, but issue #425 already involves as many aspects as I can handle at once in my mind.  I'm not able to add extra aspects to an already complex issue.  After we resolve issue #425, we can revisit your request.  At that stage of Tax-Calculator development, it will be much easier to estimate the benefits and costs of adding the ability to read in "surplus variables" and "C-values."

cc @MattHJensen @GoFroggyRun @Amy-Xu

**Comment by GoFroggyRun** [comment]
@feenberg :
I think the major issue to allow variables like `c00100` in the input file here will be that whether we should recalculate `c00100`, or skip the recalculation and carry on the inputed `c00100`. 
If we recalculate the `c00100`, then allowing it to be in the input file seems null to me. Or we can choose to carry on the inputed `c00100` without any recalculation, but then it would be troublesome: we need to avoid the recalculation, possibly by not calling the function where `c00100` will be (re)calculated. If this is the case, we might miss something else at the same time as well. (`c00100` might be easier to deal with, but in general it's problematic I imagine)
If we would like to, however, do some comparisons based on our calculated results and those inputed results (if two files share some same variable names), this shouldn't be too hard to achieve: we can put those non-valid inputs aside and have the difference outputted (if any) after the calculation is done at our end. Does this approach address part of your concerns? 

cc @martinholmer @MattHJensen

**Comment by feenberg** [comment]
On Wed, 16 Dec 2015, Sean.Wang wrote:

> 
> @feenberg :
> I think the major issue to allow variables like c00100 in the input file here will be that whether we should recalculate c00100,
> or skip the recalculation and carry on the inputed c00100.

I had not contemplated using the c00100 variable in place of our 
calculation of c00100 - I haven't had need of that. An application would 
be to set c00100 to e00100 and that would prevent errors in the AGI 
calculation from propogating further in the calculator. While an 
interesting idea, it would complicate the code.

I only thought that if c00100 were in the input file, it would be 
overwritten with the calculator version. In sas or fortran this only 
requires that c00100 be properly set before it is used. I don't know if it 
would complicate python.

dan

**Comment by martinholmer** [comment]
This issue #425 has been resolved by the merge of pull request #510 and pull request #511.
